src/Test.FunctionalTests.Sarif/PropertyBagConverterTests.cs

```diff
@@ -28,34 +27,38 @@ public void PropertyBagConverter_RoundTrip()
             string longPropertyName = "hugeFileSizeBytes";
             long longPropertyValue = (long)10 * int.MaxValue;
 
-            var run = new Run();
+            string dateTimePropertyName = nameof(dateTimePropertyName);
+            DateTime dateTimePropertyValue = new DateTime(2019, 9, 27, 13, 52, 0);
+
+            var run = new Run
+            {
+                Tool = new Tool
+                {
+                    Driver = new ToolComponent
+                    {
+                        Name = "CodeScanner"
+                    }
+                }
+            };
+
             run.SetProperty(intPropertyName, 42);
             run.SetProperty(stringPropertyName, stringPropertyValue);
             run.SetProperty(normalStringPropertyName, normalStringPropertyValue);
-            run.SetProperty<long>(longPropertyName, longPropertyValue);
+            run.SetProperty(longPropertyName, longPropertyValue);
+            run.SetProperty(dateTimePropertyName, dateTimePropertyValue);
 
             var originalLog = new SarifLog
             {
                 Runs = new List<Run>
                 {
-                    new Run
-                    {
-                        Tool = new Tool
-                        {
-                            Driver = new ToolComponent
-                            {
-                                Name = "CodeScanner"
-                            }
-                        },
-                        Properties = run.Properties
-                    }
+                    run
                 }
             };
 
             var settings = new JsonSerializerSettings { Formatting = Formatting.Indented };
             string originalLogText = JsonConvert.SerializeObject(originalLog, settings);
 
-            SarifLog deserializedLog = JsonConvert.DeserializeObject<SarifLog>(originalLogText);
+            SarifLog deserializedLog = SarifUtilities.DeserializeObject<SarifLog>(originalLogText);
             run = deserializedLog.Runs[0];
 
             int integerProperty = run.GetProperty<int>(intPropertyName);
```

Comment:
Is there a ValueComparer you can use to verify the entire Run object seems to match the Deserialized one? I think it's good to have the per-property checks for clarity, but an overall compare might be good redundant verification.

Reply:
Yes, each generated class (`Run`) has an associated generated equality comparer class (`RunEqualityComparer`) with a static `Instance` property. The generated class has a static `ValueComparer` field that is set to that instance, and a `ValueEquals` method that invokes `Equals` on the instance:

```
    public partial class Run : PropertyBagHolder, ISarifNode
    {
        public static IEqualityComparer<Run> ValueComparer => RunEqualityComparer.Instance;

        public bool ValueEquals(Run other) => ValueComparer.Equals(this, other);
```

But the equality comparison fails because of a code generation bug! When comparing the individual `SerializedPropertyInfo` objects in the property bag, the generated comparison code does an object equality comparison instead of invoking `SerializedPropertyInfo.ValueComparer.Equals`.

I filed [#1689](https://github.com/microsoft/sarif-sdk/issues/1689), "Generated code for property bag comparison is incorrect".

---
In reply to: [329663713](https://github.com/microsoft/sarif-sdk/pull/1688#discussion_r329663713) [](ancestors = 329663713)